internal/controller/etcdopstask/handler/ondemandsnapshot/ondemandsnapshot.go

```diff
@@ -0,0 +1,191 @@
+	druiderr "github.com/gardener/etcd-druid/internal/errors"
+
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// ErrEtcdNotReady represents the error in case etcd is not ready
+	ErrEtcdNotReady druidapicommon.ErrorCode = "ERR_ETCD_NOT_READY"
+	// ErrBackupNotEnabled represents the error in case backup is not enabled
+	ErrBackupNotEnabled druidapicommon.ErrorCode = "ERR_BACKUP_NOT_ENABLED"
+	// ErrCreateHTTPRequest represents the error in case of failure in creating http request
+	ErrCreateHTTPRequest druidapicommon.ErrorCode = "ERR_CREATE_HTTP_REQUEST"
+	// ErrExecuteHTTPRequest represents the error in case of failure in executing http request
+	ErrExecuteHTTPRequest druidapicommon.ErrorCode = "ERR_EXECUTE_HTTP_REQUEST"
+	// ErrCreateSnapshot represents the error in case of failure in creating snapshot
+	ErrCreateSnapshot druidapicommon.ErrorCode = "ERR_CREATE_SNAPSHOT"
+	// ErrOperationCancelled represents the error when snapshot operation is cancelled
+	ErrOperationCancelled druidapicommon.ErrorCode = "ERR_OPERATION_CANCELLED"
+)
+
+// Handler implements the task.Handler interface for handling on-demand snapshot tasks.
+type handler struct {
+	k8sClient     client.Client
+	etcdReference types.NamespacedName
+	httpClient    http.Client
+	config        druidv1alpha1.OnDemandSnapshotConfig
+}
+
+// New creates a new instance of OnDemandSnapshotTask with an optional HTTP client.
+func New(k8sClient client.Client, task *druidv1alpha1.EtcdOpsTask, httpClient *http.Client) (taskhandler.Handler, error) {
+	etcdRef := task.GetEtcdReference()
+
+	var snapshotTimeout int32
+	if task.Spec.Config.OnDemandSnapshot.Type == druidv1alpha1.OnDemandSnapshotTypeFull {
+		snapshotTimeout = *task.Spec.Config.OnDemandSnapshot.TimeoutSecondsFull
+	} else {
+		snapshotTimeout = *task.Spec.Config.OnDemandSnapshot.TimeoutSecondsDelta
+	}
+
+	return &handler{
+		k8sClient:     k8sClient,
+		etcdReference: etcdRef,
+		httpClient:    ptr.Deref(httpClient, http.Client{Timeout: time.Second * time.Duration(snapshotTimeout)}),
+		config:        *task.Spec.Config.OnDemandSnapshot,
+	}, nil
+}
+
+// Admit checks if the task can be admitted for execution.
+func (h *handler) Admit(ctx context.Context) taskhandler.Result {
+	etcd, errResult := utils.GetEtcd(ctx, h.k8sClient, h.etcdReference, druidv1alpha1.LastOperationTypeAdmit)
+	if errResult != nil {
+		return *errResult
+	}
+
+	result := h.checkPrerequisitesForSnapshot(etcd, druidv1alpha1.LastOperationTypeAdmit)
+	if result != nil {
+		return *result
+	}
+	return taskhandler.Result{
+		Description: "Admit check passed",
+		Requeue:     false,
+	}
+}
+
+// Run executes the on-demand snapshot task.
+func (h *handler) Execute(ctx context.Context) taskhandler.Result {
+	etcd, errResult := utils.GetEtcd(ctx, h.k8sClient, h.etcdReference, druidv1alpha1.LastOperationTypeExecution)
+	if errResult != nil {
+		return *errResult
+	}
+
```

Comment:
Please add a comment as to why this check is performed again, even though it wa already performed once in `Admit()`, since we want to correctly handle requeue of the `Execute()` step and account for backpu spec changes or etcd readiness changes in between the requeue.